### DISCUSSION TITLE
move_factor seems does not work

### DISCUSSION THREAD

**Discussion by HwiLu** [ClickHouse/ClickHouse#61209] [Q&A] [answered]
clickhouse version: `22.8.15`

I set default storage policy with 3 volume named ckdata, cold_volume and hot_volume ,I set `move_factor` 0.5 . The data on hot_volume already exists before I add the volume named ckdata
```
SELECT *
FROM system.storage_policies

Query id: eda4cc87-ee1e-4891-9468-391644c24e49

┌─policy_name─┬─volume_name─┬─volume_priority─┬─disks───────────────┬─volume_type─┬─max_data_part_size─┬─move_factor─┬─prefer_not_to_merge─┐
│ default     │ ckdata      │               1 │ ['ckdata']          │ JBOD        │                  0 │         0.5 │                   0 │
│ default     │ cold_volume │               2 │ ['data23']          │ JBOD        │                  0 │         0.5 │                   0 │
│ default     │ hot_volume  │               3 │ ['data21','data22'] │ JBOD        │                  0 │         0.5 │                   0 │
└─────────────┴─────────────┴─────────────────┴─────────────────────┴─────────────┴────────────────────┴─────────────┴─────────────────────┘

```

but the data in hot_volume does not  move at all, the usage of volume remained constant. 

```sql
SELECT
    *,
    free_space / total_space AS free_ratio
FROM system.disks

Query id: 1030a3e8-e3c7-4d20-8318-95dc9be89dfe

┌─name────┬─path────────────────────────────┬───free_space─┬───total_space─┬─keep_free_space─┬─type──┬─cache_path─┬─────────free_ratio─┐
│ ckdata  │ /ckdata/clickhouse/             │ 151737118720 │  238107566080 │               0 │ local │            │ 0.6372629027211129 │
│ data21  │ /data/data21/clickhouse/        │ 789753901056 │ 1889550258176 │               0 │ local │            │ 0.4179586637819079 │
│ data22  │ /data/data22/clickhouse/        │ 789753901056 │ 1889550258176 │               0 │ local │            │ 0.4179586637819079 │
│ data23  │ /data/data23/clickhouse/        │ 789753901056 │ 1889550258176 │               0 │ local │            │ 0.4179586637819079 │
│ default │ /var/lib/clickhouse/clickhouse/ │ 151737118720 │  238107566080 │               0 │ local │            │ 0.6372629027211129 │
└─────────┴─────────────────────────────────┴──────────────┴───────────────┴─────────────────┴───────┴────────────┴────────────────────┘

```

```
SELECT
    disk_name,
    count(*)
FROM system.parts
GROUP BY disk_name

Query id: b6028d9e-f0f6-4378-9fc2-6589fca2d6a5

┌─disk_name─┬─count()─┐
│ data21    │    5059 │
│ ckdata    │     281 │
│ data22    │    5046 │
└───────────┴─────────┘
```

I am not sure whether there is something wrong with my configuration, my conf like below:
```
   <storage_configuration>
        <disks>
            <ckdata>
                <path>/ckdata/clickhouse/</path>
                <type>local</type>
            </ckdata>
            <data21>
                <path>/data/data21/clickhouse/</path>
                <type>local</type>
            </data21>
            <data22>
                <path>/data/data22/clickhouse/</path>
                <type>local</type>
            </data22>
            <data23>
                <path>/data/data23/clickhouse/</path>
                <type>local</type>
            </data23>
        </disks>
        <policies>
            <default>
                <move_factor>0.5</move_factor>
                <volumes>
                    <ckdata>
                        <disk>ckdata</disk>
                    </ckdata>
                    <cold_volume>
                        <disk>data23</disk>
                    </cold_volume>
                    <hot_volume>
                        <disk>data21</disk>
                        <disk>data22</disk>
                    </hot_volume>
                </volumes>
            </default>
        </policies>
    </storage_configuration>


```

**Comment by yiyang-shao** [comment]
Hi! Is https://github.com/ClickHouse/ClickHouse/issues/48847 similar to your question?

**Reply by HwiLu** [reply]
No, I saw this issue before asking the question. It is about the balance between disks. Mine is about the balance between volumes. The introduction of move_factor refers to the balance between volumes.But I don't know why it doesn't work

**Reply by yiyang-shao** [reply]
I see. I took a deeper look. It seems that your volume priority is not correct. You might want to move `hot_volume` before `cold_volume` so that the data in `hot_volume` will be moved to `cold_volume` when the used ratio is larger than the move_factor, which I suppose is what you expect.
```xml
  <policies>
      <default>
          <move_factor>0.5</move_factor>
          <volumes>
              <ckdata>
                  <disk>ckdata</disk>
              </ckdata>
              <hot_volume>
                  <disk>data21</disk>
                  <disk>data22</disk>
              </hot_volume>
              <cold_volume>
                  <disk>data23</disk>
              </cold_volume>
          </volumes>
      </default>
  </policies>
```

**Reply by HwiLu** [reply]
Thank you for reading patiently, but it seems that you still misunderstood.

What I expect is that the data in `hot_volume ` which  space free radio is `0.4179586637819079` should move to  ckdata   which  space free radio is  `0.6372629027211129`, the volumes named ckdata has already be head and the volume_priority is smallest.

**Reply by yiyang-shao** [reply]
The smaller the volume_priority is, the "hotter" the data shall be. If you want to have `hot_volume` data move to `ckdata`, put `hot_volume` as the first volume.

**Reply by HwiLu** [reply]
😲

**Reply by HwiLu** [reply]
I tried it and it worked, so after we expand the capacity, if we want to move data, just configuring move_factor is not enough.

 Do we also need to adjust the order of the disks in the volume? It means that we still need to manually adjust the priority of the disk?
 
I always thought that priority only works when inserting and merging data.
😦

**Reply by yiyang-shao** [reply, marked as answer]
Yes. It is stated here: https://clickhouse.com/docs/en/operations/system-tables/storage_policies

> volume_priority ([UInt64](https://clickhouse.com/docs/en/sql-reference/data-types/int-uint)) — Volume order number in the configuration, the data fills the volumes according this priority, i.e. data during inserts and merges is written to volumes with a lower priority (taking into account other rules: TTL, max_data_part_size, move_factor).

But probably it is not clear enough. Volume priority (aka order of the volumes) is necessary for clickhouse to distinguish between source and destination disks during data movement processes.

**Reply by HwiLu** [reply]
This is very confusing, when insert data, data is written firstly to volumes with a lower priority , when  we want to move data out, the volume shall be a lower priority (aka smaller volume_priority ), These two situations are exactly the opposite

**Comment by HwiLu** [comment]
@den-crane Could you give me some advice? Is it a bug? How can I check it? Thanks
